Secure Controller Testing

jonathan-r-thorpe opened the pull request:
Updates required to run the NMOS Controller tests in secure mode. These include:

- Correctly forming secure URLs within the ControllerTest.py base class
- Implementing secure version of WebSocket server for mock Registry subscriptions
- Fix of bug in mock Node which could have led to spurious failure of IS-05 Controller test_02 (in both secure and insecure modes)

jonathan-r-thorpe commented:
AFAIKS this PR supersedes the modifications proposed in #688

garethsb commented on docs/2.8. Usage - Testing Controllers.md:
```diff
@@ -14,7 +14,13 @@ In the NMOS Testing Tool select one of the 'Controller' test suites and configur
-### Known Issues
+## Secure Controller Testing
+
+Before running the Controller tests in secure mode, add an entry to the hosts file for 'mocks.testsuite.nmos.tv' pointing at the IP address of the host running the test suite on the host running the test suite and the host running the Controller under test.
```
Maybe needs to correspond to config `DNS_DOMAIN`... though we don't currently make that point even in https://github.com/AMWA-TV/nmos-testing/blob/master/test_data/BCP00301/README.md#hosts-files.

jonathan-r-thorpe replied:
Hmm, yes. I guess we should get it right here though. Perhaps:

> ...add an entry to the hosts file on the host running the test suite and the host running the Controller under test pointing at the IP address of the host running the test suite. This entry should correspond to the DNS_DOMAIN and be of the form 'mocks.[DNS_DOMAIN]'.  By default this should be 'mocks.testsuite.nmos.tv'.

garethsb replied:
We should probably prefix that with "Secure testing should typically be done in a unicast DNS-SD environment, by setting `DNS_SD_MODE` (see ...)."?

jonathan-r-thorpe replied:
Documentation has now ben updated.

garethsb commented on nmos-testing-facade.py:
```diff
@@ -55,10 +55,16 @@ def index():
             valid, response = do_request('POST', form['response_url'], json=answer_json)
+
+            if not valid:
+                return response
```
This function seems to have several possible return types (response object, tuple of string and status code, or just a string). Do all of these get interpreted by the caller as intended?

jonathan-r-thorpe replied:
Yes, it's a good point, and I will check. I have a feeling that success/errors may be handled inconsistently.

jonathan-r-thorpe replied:
Responses and error handling in the Testing Facade have now been generally improved.

garethsb commented on nmostesting/Config.py:
```diff
@@ -119,6 +119,10 @@
 
+# The host name of the mocks used by the Controller tests when they are run securely. This should be consistent
+# with the hosts entries in the host running the test suite and the host running the Controller under test.
+MOCKS_HOSTNAME = "mocks.testsuite.nmos.tv"
```
Why or why not should this just be built from `DNS_DOMAIN` ?

Also, may need to be `nmos-mocks.local` or `mocks.<DNS_DOMAIN>`, depending on `DNS_SD_MODE` ?

jonathan-r-thorpe replied:
Yes good point - I will investigate `DNS_SD_MODE` configuration further.

jonathan-r-thorpe replied:
Mock hostname is now calculated based on DOMAIN_NAME and DNS_SD_MODE. It's been removed from the Config.py.

garethsb commented on nmostesting/ControllerTest.py:
```diff
@@ -79,6 +79,11 @@ class ControllerTest(GenericTest):
     def __init__(self, apis, registries, node, dns_server, disable_auto=True):
         # Remove the spec_path as there are no corresponding GitHib repos for Controller Tests
```
```suggestion
        # Remove the spec_path as there are no corresponding GitHub repos for Controller Tests
```

garethsb commented on nmostesting/ControllerTest.py:
```diff
@@ -79,6 +79,11 @@ class ControllerTest(GenericTest):
     def __init__(self, apis, registries, node, dns_server, disable_auto=True):
         # Remove the spec_path as there are no corresponding GitHib repos for Controller Tests
         apis[CONTROLLER_TEST_API_KEY].pop("spec_path", None)
+        if apis[CONTROLLER_TEST_API_KEY]["base_url"].startswith("https"):
```
Could/should this just check `ENABLE_HTTPS` ?
Or check both `"base_url"` and `"url"` separately?

jonathan-r-thorpe replied:
Yes should check `ENABLE_HTTPS` I think.  Admittedly attempting to substitute `https` for `http` will just not nothing in the `http` case, but I think the condition makes the intention explicit.

garethsb commented on nmostesting/ControllerTest.py:
```diff
@@ -117,12 +124,17 @@ def set_up_tests(self):
-        self.mock_registry_base_url = 'http://' + get_default_ip() + ':' + \
-            str(self.primary_registry.get_data().port) + '/'
-        self.mock_node_base_url = 'http://' + get_default_ip() + ':' + str(self.node.port) + '/'
+        if CONFIG.ENABLE_HTTPS:
+            self.mock_registry_base_url = 'https://' + CONFIG.MOCKS_HOSTNAME + ':' + \
+                str(self.primary_registry.get_data().port) + '/'
+            self.mock_node_base_url = 'https://' + CONFIG.MOCKS_HOSTNAME + ':' + str(self.node.port) + '/'
+        else:
+            self.mock_registry_base_url = 'http://' + get_default_ip() + ':' + \
+                str(self.primary_registry.get_data().port) + '/'
+            self.mock_node_base_url = 'http://' + get_default_ip() + ':' + str(self.node.port) + '/'
```
Seems like there may be a more elegant/maintainable version of this code struggling to get out... :-)

E.g. use `self.protocol`

jonathan-r-thorpe replied:
Yes, agreed :) Is this any better (assuming that the mocks_hosthame is calculated earlier)?
```suggestion
        host = mocks_hostname if CONFIG.ENABLE_HTTPS else get_default_ip()
        self.mock_registry_base_url = self.protocol + '://' + host + ':' + str(self.primary_registry.get_data().port) + '/'
        self.mock_node_base_url = self.protocol + '://'  + host + ':' + str(self.node.port) + '/'
```

garethsb replied:
Yep, that's the sort of thing 👍

jonathan-r-thorpe replied:
Code updated as per my comment above.

garethsb commented on nmostesting/mocks/Node.py:
```diff
@@ -505,7 +505,7 @@ def staged(version, resource, resource_id):
     # Track requests
     NODE.staged_requests.append({'method': request.method, 'resource': resource, 'resource_id': resource_id,
-                                 'data': request.json})
+                                 'data': request.get_json() if request.is_json else None})
```
According to other examples of `get_json` in this repo, and the docs at https://flask.palletsprojects.com/en/2.2.x/api/#flask.Request.get_json, this may work:

```suggestion
                                 'data': request.get_json(silent=True)})
```

jonathan-r-thorpe replied:
Yes, this works

garethsb reviewed (changes requested):
Thanks, @jonathan-r-thorpe ! Here's a quick review...

jonathan-r-thorpe commented:
BTW This PR resolves the issue raised in #668

garethsb reviewed (approved):
I haven't tested this myself but it LGTM

cristian-recoseanu commented:
Hi @jonathan-r-thorpe. Thank you for your contribution here!

I only just got round to testing this with our controller.
All working well from what I can see in my testing session.
The controller finds the secure registry and connects to the secure web socket. All tests are passing.
I've also run it in insecure mode for a bit of regression testing. All seems fine there as well.

jonathan-r-thorpe commented:
... @cristian-recoseanu hostname now present and correct in the `ws_href`.

jonathan-r-thorpe commented:
Added handling of `NMOSTestExceptions` in the test setup - this was manifesting as badly formed error messages, such as the one originally seen when attempting to do secure Controller testing, and more recently in #717.  `NMOSTestExceptions` are handled in the testing loop, but not in test setup.  The `TestResult` report is now extracted in `ControllerTest` setup and thrown as an `NMOSInitException`.

garethsb commented on nmostesting/ControllerTest.py:
```diff
@@ -117,12 +124,17 @@ def set_up_tests(self):
+        try:
+            self._populate_registry(test)
+        except NMOSTestException as e:
+            raise NMOSInitException(e.args[0].output())
```
Could this be just:
```suggestion
            raise NMOSInitException(e.args[0])
```

garethsb reviewed (approved):
LGTM

jonathan-r-thorpe commented:
Thanks @garethsb and @cristian-recoseanu ! Yes, @garethsb I think there are still issues with multiple network adapters - I'm still confused by that particular problem as the mock resources bind to '0.0.0.0' which (I thought) meant they were available on *all* network adapters - but obviously not in the case highlighted. Yes, please merge :)

garethsb commented:
> '0.0.0.0' which (I thought) meant they were available on all network adapters - but obviously not in the case highlighted.

I think that too! to be investigated......
